art/attacks/evasion/wasserstein.py

```diff
@@ -0,0 +1,1151 @@
+                  samples. Otherwise, model predictions are used as labels to avoid the "label leaking" effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236). Default is `None`.
+        :type y: `np.ndarray`
+        :param cost_matrix: A non-negative cost matrix.
+        :type cost_matrix: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        y = check_and_transform_label_format(y, self.estimator.nb_classes)
+        x_adv = x.copy().astype(ART_NUMPY_DTYPE)
+
+        if y is None:
+            # Throw error if attack is targeted, but no targets are provided
+            if self.targeted:
+                raise ValueError("Target labels `y` need to be provided for a targeted attack.")
+
+            # Use model predictions as correct outputs
+            targets = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
+        else:
+            targets = y
+
+        # Compute the cost matrix if needed
+        cost_matrix = kwargs.get("cost_matrix")
+        if cost_matrix is None:
+            cost_matrix = self._compute_cost_matrix(self.p, self.kernel_size)
+
+        # Compute perturbation with implicit batching
+        nb_batches = int(np.ceil(x.shape[0] / float(self.batch_size)))
+        for batch_id in range(nb_batches):
+            logger.debug("Processing batch %i out of %i", batch_id, nb_batches)
+
+            batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+            batch = x_adv[batch_index_1: batch_index_2]
+            batch_labels = targets[batch_index_1: batch_index_2]
+
+            x_adv[batch_index_1:batch_index_2] = self._generate_batch(
+                batch,
+                batch_labels,
+                cost_matrix,
+                self.kernel_size,
+                self.max_iter,
+                self.norm,
+                self.ball,
+                self.targeted,
+                self.eps,
+                self.eps_iter,
+                self.eps_factor,
+                self.eps_step,
+                self.regularization,
+                self.conjugate_sinkhorn_max_iter,
+                self.projected_sinkhorn_max_iter,
+                batch.shape[0],
+            )
+
+        logger.info(
+            "Success rate of attack: %.2f%%",
+            100 * compute_success(self.estimator, x, y, x_adv, self.targeted, batch_size=self.batch_size),
+        )
+
+        return x_adv
+
+    def _generate_batch(
+            self,
+            x,
+            targets,
+            cost_matrix,
+            kernel_size,
+            max_iter,
+            norm,
+            ball,
+            targeted,
+            eps,
+            eps_iter,
+            eps_factor,
+            eps_step,
+            regularization,
+            conjugate_sinkhorn_max_iter,
+            projected_sinkhorn_max_iter,
```

Comment:
Are these arguments required? Could they be replaced by direct access to self inside self._generate_batch?

Reply:
See comments above.